exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/trace_exporter/__init__.py

```diff
@@ -0,0 +1,185 @@
+
+
+DEFAULT_COMPRESSION = Compression.NoCompression
+DEFAULT_ENDPOINT = "http://localhost:55681/v1/traces"
+DEFAULT_TIMEOUT = 10  # in seconds
+
+
+class OTLPSpanExporter(SpanExporter):
+
+    _MAX_RETRY_TIMEOUT = 64
+
+    def __init__(
+        self,
+        endpoint: Optional[str] = None,
+        certificate_file: Optional[str] = None,
+        headers: Optional[Dict[str, str]] = None,
+        timeout: Optional[int] = None,
+        compression: Optional[Compression] = None,
+    ):
+        self._endpoint = endpoint or environ.get(
+            OTEL_EXPORTER_OTLP_TRACES_ENDPOINT,
+            environ.get(OTEL_EXPORTER_OTLP_ENDPOINT, DEFAULT_ENDPOINT),
+        )
+        self._certificate_file = certificate_file or environ.get(
+            OTEL_EXPORTER_OTLP_TRACES_CERTIFICATE,
+            environ.get(OTEL_EXPORTER_OTLP_CERTIFICATE, False),
+        )
+        self._headers = headers or _headers_from_env()
+        self._timeout = timeout or int(
+            environ.get(
+                OTEL_EXPORTER_OTLP_TRACES_TIMEOUT,
+                environ.get(OTEL_EXPORTER_OTLP_TIMEOUT, DEFAULT_TIMEOUT),
+            )
+        )
+        self._compression = compression or _compression_from_env()
+        self._session = requests.Session()
+        self._session.headers.update(self._headers)
+        self._session.headers.update(
+            {"Content-Type": ProtobufEncoder._CONTENT_TYPE}
+        )
+        if self._compression is not Compression.NoCompression:
+            self._session.headers.update(
+                {"Content-Encoding": self._compression.value}
+            )
+        self._shutdown = False
+
+    def _export(self, serialized_data: str):
+        data = serialized_data
+        if self._compression == Compression.Gzip:
+            gzip_data = BytesIO()
+            with gzip.GzipFile(fileobj=gzip_data, mode="w") as gzip_stream:
+                gzip_stream.write(serialized_data)
+            data = gzip_data.getvalue()
+        elif self._compression == Compression.Deflate:
+            data = zlib.compress(bytes(serialized_data))
+
+        return self._session.post(
+            url=self._endpoint,
+            data=data,
+            verify=self._certificate_file,
+            timeout=self._timeout,
+        )
+
+    @staticmethod
+    def _retryable(resp: requests.Response) -> bool:
+        if resp.status_code == 408:
+            return True
+        if resp.status_code >= 500 and resp.status_code <= 599:
+            return True
+        return False
+
+    def export(self, spans) -> SpanExportResult:
+        # After the call to Shutdown subsequent calls to Export are
+        # not allowed and should return a Failure result.
+        if self._shutdown:
+            _logger.warning("Exporter already shutdown, ignoring batch")
+            return SpanExportResult.FAILURE
+
+        serialized_data = ProtobufEncoder.serialize(spans)
+
+        for delay in expo(max_value=self._MAX_RETRY_TIMEOUT):
```

Comment:
Could this code could be refactored to use backoff's `on_exception` decorator? Not sure if it would work or not, but it may simplify the code a bit.

Reply:
Retry here is conditionally based on the status code. I couldn't find how do we achieve that using `on_exception`.

Reply:
The following code seems to work:
```
import backoff
import requests

def fatal_code(e):
    if e.response.status_code == 408 or 500 <= e.response.status_code <= 599:
        return False
    return True

@backoff.on_exception(backoff.expo,
                      requests.exceptions.RequestException,
                      max_time=10,
                      giveup=fatal_code)
def get_url(url):
    r = requests.get(url)
    r.raise_for_status()
    return r

resp = get_url("http://localhost:8080")
print(resp)
```

Note that this requires calling `raise_for_status` to be called on the response object for it to work at all.

Reply:
to avoid calling `ProtobufEncoder.serialize` again and again this needed another indirection call and required `raise_for_status` and some other stuff which made it look code more complicated than it should be. If you feel strongly about using the `on_exception` decorator I will change otherwise it seems to me not worth it.

Reply:
I think this is fine as is.